src/sentry/static/sentry/app/components/forms/form.tsx

```diff
@@ -112,24 +112,10 @@ class Form<
   };
 
   onSubmitSuccess = (data: object) => {
-    const curData = this.state.data;
-    let newData = {};
-    if (data) {
-      Object.keys(curData).forEach(k => {
-        if (data.hasOwnProperty(k)) {
-          newData[k] = data[k];
-        } else {
-          newData[k] = curData[k];
-        }
-      });
-    } else {
-      newData = curData;
-    }
-
     this.setState({
       state: FormState.READY,
       errors: {},
-      initialData: newData,
+      initialData: {...this.state.data, ...(data || {})},
```

Comment:
`data` should never be `null` or `undefined`, right?

Reply:
`null` and `undefined` would be fine, since those are both falsey. Right?

```js
> null || {}
{}
> undefined || {}
{}
>
```

Something like a non-empty string is what would break this:

```
> {...{}, ...{}}
{}
> {...{}, ..."cheese"}
{ '0': 'c', '1': 'h', '2': 'e', '3': 'e', '4': 's', '5': 'e' }
>
```

P.S. Interestingly, empty string would pass unnoticed. :)

```
> {...{}, ...""}
{}
```

Reply:
Ah, sorry for not being clearer. I was questioning the need for the empty object fallback. If this can never be `null` or `undefined`, why do we even need the `|| {}` part?

Reply:
Ah, gotcha. I was preserving the `if (data)` guard from the original. I'm kinda nervous about tacking on this refactor since it's not necessary to fix #21159, and introduces a regression risk. Therefore I tried hard to preserve the original logic. 😬